(#198) Added "Recommended package parameters" page (#203)
chocolatey/docs: input/en-us/create/recommended-package-parameters.md

---
Order: 25
xref: recommended-package-parameters
Title: Recommended Package Parameters
Description: The standart of naming package parameters and the list of most commonly used ones.
---

Chocolatey's package parameters provide a way for package users to configure how the package is installed and configured.

As the parameters are handled individualy by each package, establishing specific parameter names for common actions significantly simpifies the life of a user.

## Naming convention
* Parameter names should be in CamelCase
* Disabling parameters should start with 'No' (`/NoStartup`, `/NoFileAssociation`, etc.)
* Binary actions should have both Off (`/NoDesktopShortcut`) and On (`/DesktopShortcut`) parameters, overriding default behaviour.
* In cases where there are multiple binary actions of the same type (e.g. multiple components, desktop shortcuts) they should be separated by a comma (`/NoShellIntegration:GuiHere,ShellHere`, `/NoDesktopShortcut:AllUsers,CurrentUser`) 

> * Actions expanding on exsting parameter behaviour should create subparameters, not create new ones (e.g. `/DesktopShortcut:Component1Shortcut,Component2Shortcut`, not `/DesktopShortcutComponent1 /DesktopShortcutComponent2`)

Reviewer: I agree. Having an explicit parameter for each component feels like it would be much clearer for the user, instead of a single parameter accepting each component.

My thoughts are that arguments like `/NoDesktopShortcut`, `/DesktopShortcut` (and other similar parameters) should disable/enable for all components while if there is a need to only make the change for certain components those are better specified as separate parameters.

> * In cases where there are multiple binary actions of the same type (e.g. multiple components, desktop shortcuts) they should be separated by a comma (`/NoShellIntegration:GuiHere,ShellHere`, `/NoDesktopShortcut:AllUsers,CurrentUser`) 
> * Actions expanding on exsting parameter behaviour should create subparameters, not create new ones (e.g. `/DesktopShortcut:Component1Shortcut,Component2Shortcut`, not `/DesktopShortcutComponent1 /DesktopShortcutComponent2`)

Author: ```suggestion
* In cases where there are multiple binary actions of the same type (like multiple components, desktop shortcuts) there should be a single main action (e.g. /NoDesktopShortcut), which when passed will enable/disable every subactions, and all the subactions, created using the component name preceded by the main action (Subactions: `/NoDesktopShortcutComponent1`, `/DesktopShortcutComponent2`,)
```

> * Actions expanding on exsting parameter behaviour should create subparameters, not create new ones (e.g. `/DesktopShortcut:Component1Shortcut,Component2Shortcut`, not `/DesktopShortcutComponent1 /DesktopShortcutComponent2`)

Reviewer 2: Can you rewrite the above as it's very confusing. It's needs to be clear and simple or people will not follow it.

## Package Parameters
> :memo: **NOTE** Not all of these parameters are required to be checked by the package script. 

> * `/InstallDir:` - Changes package's installation directory.

Reviewer 2: ```suggestion
* `/InstallDir:` - Changes package's installation directory. Note that this can also be done using the `--install-arguments` switch if the installer supports it.

```

Author: I don't think `--install-arguments ` should be mentioned here, as the `/InstallDir` package parameter should replace the installer dependant arguments in the first place.

Reviewer 2: I disagree. I think `/InstallDir` would really be for when you have to do some magic inside the package to override  the installation directory, or the installer uses a parameter to indicate the install directory that `--install-arguments` doesn't know about / understand.

Author: Using --install-arguments would defeat the purpose of creating standardized package parameters. Because --install-arguments pass parameters directly to the setup program, the user must check what particular installer the packages uses (to specify the correct arguments). Package parameters instead provide a installer agnostic way of specifying options.

Reviewer 2: I feel we are going off track here. Remember, the goal of this page is to help the user by providing a set of _suggested_ package parameters that a maintainer could use. What I'm suggesting we do is tell the user another way to do the thing that the `/InstallDir` package parameter does. We're not forcing the user to check anything. We're giving them options. I think that's reasonable.

`--install-arguments` is a feature of Chocolatey that has been around for some time. The impression I am getting is that you feel `/InstallDir` would replace that, or should be used instead of that. Just to be clear, that's not the direction we want to go in here.

> * `/NoDesktopShortcut` and `/DesktopShortcut` - Explicitly set whether the package should or should not create a shortcut on desktop, if not specified the installers default parameter is used.

Reviewer 2: ```suggestion
* `/NoDesktopShortcut` and `/DesktopShortcut` - Explicitly set whether the package should or should not create a shortcut on desktop. If not specified, the installer default behaviour will apply.
```

> * `/NoStartMenuShortcut` and `/StartMenuShortcut` - Explicitly set whether the package should or should not create an entry in start menu.

Reviewer 2: ```suggestion
* `/NoStartMenuShortcut` and `/StartMenuShortcut` - Explicitly set whether the package should or should not create an entry in start menu. If not specified, the installer default behaviour will apply.
```

> * `/NoStartup` and `/Startup` - Explicitly set whether the packaged application should or should not launch on start up(Applicable to installers that already can startup by default).

Reviewer 2: ```suggestion
* `/NoStartup` and `/Startup` - Explicitly set whether the packaged application should or should not launch on startup, if applicable. If not specified, the installer default behaviour will apply.
```

> * `/NoQuickLaunchShortcut` and `/QuickLaunchShortcut` - Explicitly set whether the package should or should not create a shortcut in quick launch menu.

Reviewer 2: ```suggestion
* `/NoQuickLaunchShortcut` and `/QuickLaunchShortcut` - Explicitly set whether the package should or should not create a shortcut in quick launch menu. If not specified, the installer default behaviour will apply.
```

> * `/NoFileAssociation` and `/FileAssociation` - Whether to associate files with this application.

Reviewer 2: ```suggestion
* `/NoFileAssociation` and `/FileAssociation` - Whether to associate files with this application. If not specified, the installer default behaviour will apply.
```

> * `/NoShellIntegration:` and `/ShellIntegration:` - Explicitly set whether the package should or should not add itself into explorer's context menu

Reviewer 2: ```suggestion
* `/NoShellIntegration:` and `/ShellIntegration:` - Explicitly set whether the package should or should not add itself into explorer's context menu. If not specified, the installer default behaviour will apply.
```

Author: ```suggestion
* `/NoShellIntegration` and `/ShellIntegration` - Explicitly set whether the package should or should not add itself into explorer's context menu
```
Colon removed, as /ShellIntegration is a boolen.

> * `/NoAppendPath` and `/AppendPath` - Add program folder directory to environment variable "PATH"

Reviewer 2: ```suggestion
* `/NoAppendPath` and `/AppendPath` - Add program folder directory to environment variable "PATH". If not specified, the installer default behaviour will apply.
```

> * `/Components:` - override default installation components (every optional component is turned off, and allowed selectively), e.g.: `/Components:Include_pip,CompileAll`. Package maintainer must mention the default components list.

Reviewer 2: @gep13 @AdmiringWorm This is a follow-on from the comments I made above.

Reviewer 2: Can this be rewritten in the same way as [this one](https://github.com/chocolatey/docs/pull/203/files#r740441142).

> * `/Language:` - Specifies the language to use in the form of Language code (ISO 639-1), Hyphen (-) and Country code (ISO 3166-1 alpha-2), e.g.: en-US, en-UK

Reviewer 2: Some software will not use the language / country code format. Maintainers may therefore feel they can't use this parameter. I'd amend this to ensure that it uses the language codes that are used by the installer so that they don't feel they have to convert from what is above to what the installer uses.

Reviewer 2: That's a fair point. But I think you have to balance it with what is convenient for the maintainer too.

If we have a whole bunch of package parameters that are a pain for the maintainer to implement, they simply won't implement them, (as a maintainer, I wouldn't) and that is a detriment to the user. Remember, the package parameters on this page are _suggestions_. Lowering the friction to using them is a goal for both users and maintainers.

So if we leave it for the maintainer to decide, then the users only has to read the package description where the package parameters are described. The users will have to do that anyway to know what package parameters are available for them to use. So, it's not additional work.

* `/LicenseUsername:`, `/LicenseKey:` and `/LicensePath:` - Registration information required for package installation.

> * `/Username:` and `/Password:` - Parameters to allow login into the remote download server for the package or the program, it should be allowed for the password to be read from the prompt.

Reviewer 2: I think these parameters should be a bit more loose in their definition. What if it's a username and password to set for the application to open. Or something else?

What should happen here if the user does not input the information at the prompt / it times out?

* `/Purge` - Remove user files before install, or after uninstall.

## Package Example
### Package Description
~~~markdown
# Package Parameters
The following package parameters can be set:

>  * `/InstallDir:` - Where to install the binaries to, Default: `$env:ProgramFiles\GreatInnoProgram`

Reviewer 2: ```suggestion
 * `/InstallDir:` - Where to install the binaries to. Default is `$env:ProgramFiles\GreatInnoProgram`. Example `/InstallDir:C:\Tools\MyProgram`.
```

>  * `/Components:` - Configure what additional components to install, Possible values: `docs`, `examples`, `sources`, Default: `/Components:docs,examples` 

Reviewer 2: @gep13 @AdmiringWorm Follow-on from the above again.

Reviewer 2: Again, same as [above](https://github.com/chocolatey/docs/pull/203/files#r740441142).

>  * `/NoStartMenuShortcut:` and `/StartMenuShortcut:` - Whether to add an entry into the "All Programs" folder, Default: Yes

Reviewer 2: ```suggestion
 * `/NoStartMenuShortcut` and `/StartMenuShortcut` - Whether or not to add an entry into the "All Programs" folder.
```

>  * `/NoDesktopShortcut` and `/DesktopShortcut` - Whether to add a shortcut on desktop, Default: No

Reviewer 2: ```suggestion
 * `/NoDesktopShortcut` and `/DesktopShortcut` - Whether or not to add a shortcut on desktop.
```

>  * `/NoFileAssociation` and `/FileAssociation` - Whether to associate applicable files with this program, Default: No

Reviewer 2: ```suggestion
 * `/NoFileAssociation` and `/FileAssociation` - Whether or not to associate applicable files with the installed software.
```

>  * `/NoAppendPath` and `/AppendPath` - Whether to add this programm's folder to 'PATH' environmental variable, Default: Yes

Reviewer 2: ```suggestion
 * `/NoAppendPath` and `/AppendPath` - Whether or not to add this software install folder to 'PATH' environmental variable. 
```

>  * `/NoShellIntegration` and `/ShellIntegration` - Whether the package should or should not add itself into explorer's context menu, Subparameters: `addcontextmenufiles` - add to file' context menu, `addcontextmenufolders` - add to folder' context menu, Default: `/NoShellIntegration:addcontextmenufiles /ShellIntegration:addcontextmenufolders`

Reviewer 2: @gep13 @AdmiringWorm ... you know the drill 😄

Reviewer 2: Again, same as [above](https://github.com/chocolatey/docs/pull/203/files#r740441142).

To pass parameters, use `--params "''"` (e.g. `choco install packageID [other options] --params="'/Item1:value /Item2:value2 /FlagBoolean'"`).

> To have choco remember parameters on upgrade, be sure to set `choco feature enable -n=useRememberedArgumentsForUpgrades`.

Reviewer 2: ```suggestion
To remember the parameters you have set when you upgrade a package, set `choco feature enable -n=useRememberedArgumentsForUpgrades`.
```

~~~
### ChocolateyInstall.ps1

> ~~~powershell

Reviewer: @pauby should this example be noted to be very specific about Inno Setup installers?

If someone tried to use the same example for other technologies it would not work.
